File: zig/src/day2.zig
const std = @import("std");
const util = @import("util.zig");

pub fn part1(input: []const u8) u64 {
    var pos: u64 = 0;
    var depth: u64 = 0;

    var lines = std.mem.split(u8, input, "\n");
    while (lines.next()) |line| {
        if (line.len == 0) continue;

        var parts = std.mem.split(u8, line, " ");
        const command = parts.next().?;
        const distance = util.parseInt(u64, parts.next().?);

        if (std.mem.eql(u8, command, "forward")) {
            pos += distance;
        } else if (std.mem.eql(u8, command, "up")) {
            depth -= distance;
        } else if (std.mem.eql(u8, command, "down")) {
            depth += distance;
        }
    }

    return pos * depth;
}

pub fn part2(input: []const u8) u64 {
    var pos: u64 = 0;
    var depth: u64 = 0;
    var aim: u64 = 0;

    var lines = std.mem.split(u8, input, "\n");
    while (lines.next()) |line| {
        if (line.len == 0) continue;

        var parts = std.mem.split(u8, line, " ");
        const command = parts.next().?;
        const distance = util.parseInt(u64, parts.next().?);

        if (std.mem.eql(u8, command, "forward")) {
            pos += distance;
            depth += aim * distance;
        } else if (std.mem.eql(u8, command, "up")) {
            aim -= distance;
        } else if (std.mem.eql(u8, command, "down")) {
            aim += distance;
        }
    }

    return pos * depth;
}

test "day 2 part 1" {
    const test_input =
        \\forward 5
        \\down 5
        \\forward 8
        \\up 3
        \\down 8
        \\forward 2
    ;

    try std.testing.expectEqual(part1(test_input), 150);
}

test "day 2 part 2" {
    const test_input =
        \\forward 5
        \\down 5
        \\forward 8
        \\up 3
        \\down 8
        \\forward 2
    ;

    try std.testing.expectEqual(part2(test_input), 900);
}

File: rust/src/day2.rs
enum Command {
    Forward(i32),
    Down(i32),
    Up(i32),
}

fn parse_input(input: &'_ str) -> impl Iterator<Item = Command> + '_ {
    input.split('\n').filter(|s| !s.trim().is_empty()).map(|s| {
        let mut parts = s.split(' ');
        let cmd = parts
            .next()
            .expect("each command should have a command word");
        let dis = parts
            .next()
            .expect("each command should have a distance")
            .parse()
            .expect("each command should have a distance following it");
        match cmd {
            "forward" => Command::Forward(dis),
            "down" => Command::Down(dis),
            "up" => Command::Up(dis),
            _ => panic!("invalid command"),
        }
    })
}

pub fn part1(input: &str) -> i32 {
    let (hor, depth) = parse_input(input).fold((0, 0), |(hor, depth), cmd| match cmd {
        Command::Forward(dis) => (hor + dis, depth),
        Command::Down(dis) => (hor, depth + dis),
        Command::Up(dis) => (hor, depth - dis),
    });
    hor * depth
}

pub fn part2(input: &str) -> i32 {
    let (hor, depth, _) = parse_input(input).fold((0, 0, 0), |(hor, depth, aim), cmd| match cmd {
        Command::Forward(dis) => (hor + dis, depth + dis * aim, aim),
        Command::Down(dis) => (hor, depth, aim + dis),
        Command::Up(dis) => (hor, depth, aim - dis),
    });
    hor * depth
}

#[cfg(test)]
mod tests {
    use super::*;

    const TEST_INPUT: &str = r#"
forward 5
down 5
forward 8
up 3
down 8
forward 2
    "#;

    #[test]
    fn test_part1() {
        assert_eq!(part1(TEST_INPUT), 150);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2(TEST_INPUT), 900);
    }
}
